File: leetcode/problems/language-specific/go/6.zigzag-conversion.go
/*
 * @lc app=leetcode id=6 lang=golang
 *
 * [6] Zigzag Conversion
 */

// @lc code=start
func convert(s string, numRows int) string {
	if numRows == 1 {
		return s
	}

	res, rowSlice, curRow, goingDown := "", make([]string, findMin(numRows, len(s))), 0, false
	for _, v := range s {
		rowSlice[curRow] += string(v)
		if curRow == 0 || curRow == numRows-1 {
			goingDown = !goingDown
		}
		if goingDown {
			curRow++
		} else {
			curRow--
		}
	}

	for _, v := range rowSlice {
		res += v
	}

	return res
}

func findMin(a, b int) int {
	if a <= b {
		return a
	}

	return b
}

// @lc code=end

File: leetcode/problems/problem-specific/6.zigzag-conversion/6.zigzag-conversion.py
#
# @lc app=leetcode id=6 lang=python3
#
# [6] Zigzag Conversion
#

# @lc code=start

# * Iterative Solution | O(n) Time | O(n) Space
# * n -> The length of s string


class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1:
            return s

        res, rows_val = [], [[] for _ in range((min(numRows, len(s))))]
        cur_row, is_going_down = 0, False
        for char in s:
            rows_val[cur_row].append(char)
            if cur_row == 0 or cur_row == numRows - 1:
                is_going_down = not is_going_down
            if is_going_down:
                cur_row += 1
            else:
                cur_row -= 1

        for row_val in rows_val:
            res.append("".join(row_val))

        return "".join(res)


# @lc code=end
